### DISCUSSION TITLE
Using CriticMarkup in quarto document - colouring the critic markups?

### DISCUSSION THREAD

**Discussion by rkrug** [quarto-dev/quarto-cli#449] [Q&A]
Hi

I am getting into quarto and like it a lot. Thanks.

I am using [critic markup](https://github.com/CriticMarkup/CriticMarkup-toolkit) to mark comments etc in the markdown document.

From the website above:

> There are five types of Critic marks:

> Addition {++ ++}
> Deletion {-- --}
> Substitution {~~ ~> ~~}
> Comment {>> <<}
> Highlight {== ==}{>> <<}
> Using these five basic marks you can successfully copy edit in plain text.

 Is there a way (pre-processor?) to colour these in the output document? It would be much easier to see them in the compiled document?

Thanks

**Comment by jjallaire** [comment]
There are a few critic markup pre-processors for pandoc out there. I personally think that the best way to handle this is via a filter that adds spans (like this one: https://gist.github.com/sabgaby/988c1467a1ee28f0cdd2f09f8d26cf57) and then can be styled as appropriate. Ideally this filter would be written in Lua rather than Python so that its easy to install/use in more environments.

**Reply by rkrug** [reply]
Sounds interesting. How can I use these with quarto? Filter or pre-processor - no preferences at all, as long as it works nicely "out of the box" (mostly...).

**Comment by jjallaire** [comment]
Filters are straightforward to use w/ Quarto: https://quarto.org/docs/authoring/shortcodes-and-filters.html#filters

For a Python filter you would do this in Quarto config:

```yaml
filters:
  - type: json
    path: filter.py
```

There isn't much to the Python filter I sent a link to, so if you are super motivated to make this work well for other users I'd suggest porting it to Lua (https://pandoc.org/lua-filters.html)

**Reply by rkrug** [reply]
OK - I have saved the gist you posted and added

```
filters:
  - type: json
    path: criticpandoc.py
```

but I get the following error when running `quarto preview manuscript.Rmd`:


```
ERROR: TypeError: path62.replace is not a function
```

**Comment by jjallaire** [comment]
I wish I had time to debug this in more depth but I sadly don't. While those links should give you the general direction of the approach to solving the problem, they don't appear to be a complete, running, out of the box solution. If you have the time and inclination I'd suggest studying the code and implementing a Lua version of it.

**Reply by rkrug** [reply]
Thanks - maybe later. At the moment, I have no time for that.

**Comment by RoyiAvital** [comment]
@rkrug , Did you find any solution to that?

**Comment by mloubout** [comment]
Not sure if still looking for a solution but I made a quarto filter for it

https://github.com/mloubout/critic-markup

And what the rendering looks like

https://mloubout.github.io/critic-markup/

**Reply by RoyiAvital** [reply]
This is great! Thank you...

**Reply by rkrug** [reply]
Looks great - thanks a lot!

**Reply by harrelfe** [reply]
Fantastic work @mloubout .  Does that approach work when rendering to Word .docx or pdf files?  Or is it html-output-specific?

**Reply by mloubout** [reply]
Currently this is html specific, I hadn't had time to work on it much but the goal ultimately is to support pdf/docx with the option to choose to render original or edited
Rendering the version with markup might take a bit more work.

**Reply by harrelfe** [reply]
Thanks very much.  It would be terrific to have a format-agnostic tool.  Looking farther down the road it would be nice to define a programmable track changes markup.  Programmable in the sense that you could define a series of markups that you can turn on and off at the top of the script.  For example if track changes of type K is turned off at the top, the tracking is not rendered and only the final text shows.  I guess if this got too complicated it could possibly be better to render this like `r ch('old text', 'new text', chK)` where `chK` is set to 0 or 1 early in the script.  Just thinking out loud.
